ugvc/pipelines/lpr/train_lib_prep_recalibration_model.py

```diff
@@ -0,0 +1,114 @@
+
+    tp_args = [f"{out_dir}/estimated_tps.{suffix}" for suffix in ("bed", "vcf.gz", "parquet")]
+    fp_args = [f"{out_dir}/estimated_fps.{suffix}" for suffix in ("bed", "vcf.gz", "parquet")]
+    tp_args.extend([args.tp_min_af, 1])
+    fp_args.extend([0, args.fp_max_af])
+    afs_bed = f"{out_dir}/afs.bed"
+
+    if args.calls_vcf:
+        calls_vcf = f"{out_dir}/calls.snps.pass.vcf.gz"
+        featuremap_on_calls = f"{out_dir}/featuremap_on_calls.vcf.gz"
+        sp.print_and_run(f"bcftools view --type snps -f PASS {unfiltered_calls} -Oz -o {calls_vcf}")
+        sp.print_and_run(
+            f"bedtools intersect -a {featuremap_vcf} -b {calls_vcf} -header | "
+            f"bcftools view -Oz -o {featuremap_on_calls}"
+        )
+        featuremap_vcf = featuremap_on_calls
+        sp.print_and_run(
+            f"bcftools query -f '%CHROM %POS [%VAF]\\n' {calls_vcf} | grep -v , | "
+            + "awk -v OFS='\t' '{print $1,$2-1,$2,$3}' "
+            + f"> {afs_bed}"
+        )
+    else:
+        sp.print_and_run(
+            "bcftools query -f '%CHROM %POS %X_READ_COUNT' |awk -v OFS='\t' '{print $1,$2-1,$2,$3}' | "
+            " | bedtools groupby -c 3 -full -o count | awk -v OFS='\t' '{print $1,$2,$3,$5/$4}' " + f" > {afs_bed}"
+        )
+
+    for bed_file, vcf_file, parquet_file, min_af, max_af in (tp_args, fp_args):
+        sp.print_and_run(f"awk '$4 >= {min_af} && $4 <= {max_af}' {afs_bed} > {bed_file}")
+        sp.print_and_run(
+            f"bedtools intersect -a {featuremap_vcf} -b {bed_file} -u -header | " f"bcftools view -Oz -o {vcf_file}"
+        )
+        sp.print_and_run_clt(
+            featuremap_to_dataframe.run, [], {"-i": vcf_file, "-r": args.ref_fasta, "-f": "TGCA", "-o": parquet_file}
+        )
+
+    base_dir = args.notebooks_dir
+    balance_motifs = "True" if args.balance_motifs else "False"
+    balance_tp_fp = "True" if args.balance_tp_fp else "False"
+    output_suffix_arg = args.output_suffix if args.output_suffix != "" else '""'
+    sp.print_and_run(
+        f"papermill {base_dir}/train_lib_prep_xgboost.ipynb "
+        f"{out_dir}/train_lib_prep_xgboost{args.output_suffix}.nbconvert.ipynb "
+        f"-p work_dir {out_dir} "
+        f"-p suffix {output_suffix_arg} "
+        f"-p balance_motifs {balance_motifs} "
+        f"-p balance_tp_fp {balance_tp_fp} "
+    )
+
+    sp.print_and_run(
+        "jupyter nbconvert --to html "
+        f"{out_dir}/train_lib_prep_xgboost{args.output_suffix}.nbconvert.ipynb --template full "
+        f"--no-input --output train_lib_prep_xgboost_report{args.output_suffix}.html"
+    )
+
+
+def localize_inputs(args, sp):
+    if args.substitution_analysis_wid:
+        featuremap_vcf = (
+            f"gs://cromwell-backend-ultima-data-307918/cromwell-execution/SubstitutionAnalysis/"
+            f"{args.substitution_analysis_wid}/call-FeatureMap/FeatureMap/*/"
+            f"call-FeatureMapMerge/*.featuremap.vcf.gz"
+        )
```

Comment:
I usually do something like this:
`
import json

with open(cloud_sync(f"gs://cromwell-backend-ultima-data-307918/cromwell-execution/SubstitutionAnalysis/{args.substitution_analysis_wid}/metadata.json")) as fh:
    metadata = json.load(fh)
featuremap_vcf = metadata["outputs"]["featuremap"]
`